consistent_df/enforce_schema.py

```
@@ -0,0 +1,122 @@
+    - an empty DataFrame: Adds missing required and optional columns.
+    - a DataFrame with one or more columns: Adds missing optional columns, and
+      raises an exception if any required columns are missing.
+
+    Existing required and optional columns are converted to specified dtypes,
+    leading to an error if type conversion fails.
+
+    Args:
+        data (pd.DataFrame | None): The DataFrame to validate and transform.
+        schema (pd.DataFrame): A DataFrame defining the schema. It must include:
+            - 'column_name': Name of the column.
+            - 'dtype': Expected data type (e.g., 'int64', 'float64', 'datetime64[ns]').
+            - 'mandatory': Boolean flag indicating if the column
+                           is required (True) or optional (False).
+        keep_extra_columns (bool): If True, columns not listed in the schema are retained.
+                                   If False, only columns defined in the schema will be kept.
+
+    Returns:
+        pd.DataFrame: A DataFrame with type-consistent columns and order.
+
+    Raises:
+        ValueError: If the input DataFrame is not empty and is missing any
+                    required columns.
+        TypeError: If the input data is not a pandas DataFrame or None, or
+                   if dtype conversion fails due to incompatible types.
+        ValueError: If the schema has missing or invalid dtype values.
+
+    Example:
+        >>> PRICE_SCHEMA_CSV = \"\"\"
+        ... column_name,        dtype,                mandatory
+        ... ticker,             string[python],       True
+        ... date,               datetime64[ns],       True
+        ... currency,           string[python],       True
+        ... price,              Float64,              True
+        ... \"\"\"
+        >>> PRICE_SCHEMA = pd.read_csv(StringIO(PRICE_SCHEMA_CSV), skipinitialspace=True)
+        >>> df = pd.DataFrame({'ticker': ['AAPL', 'GOOGL'], 'price': [150.0, 2800.0]})
```

This example should IMO raise an error (missing required columns).